core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportRefreshAction.java

```diff
@@ -46,74 +51,91 @@
 /**
  * Refresh action.
  */
-public class TransportRefreshAction extends TransportBroadcastAction<RefreshRequest, RefreshResponse, ShardRefreshRequest, ShardRefreshResponse> {
+public class TransportRefreshAction extends HandledTransportAction<RefreshRequest, RefreshResponse> {
 
-    private final IndicesService indicesService;
+    private final TransportReplicatedRefreshAction replicatedRefreshAction;
+    ClusterService clusterService;
 
     @Inject
     public TransportRefreshAction(Settings settings, ThreadPool threadPool, ClusterService clusterService,
-                                  TransportService transportService, IndicesService indicesService,
-                                  ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
-        super(settings, RefreshAction.NAME, threadPool, clusterService, transportService, actionFilters, indexNameExpressionResolver,
-                RefreshRequest.class, ShardRefreshRequest.class, ThreadPool.Names.REFRESH);
-        this.indicesService = indicesService;
+                                  TransportService transportService,
+                                  ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver, TransportReplicatedRefreshAction replicatedRefreshAction) {
+        super(settings, RefreshAction.NAME, threadPool, transportService, actionFilters, indexNameExpressionResolver, RefreshRequest.class);
+        this.replicatedRefreshAction = replicatedRefreshAction;
+        this.clusterService = clusterService;
     }
 
     @Override
-    protected RefreshResponse newResponse(RefreshRequest request, AtomicReferenceArray shardsResponses, ClusterState clusterState) {
+    protected void doExecute(final RefreshRequest request, final ActionListener<RefreshResponse> listener) {
+        logger.trace("execute refresh");
+        GroupShardsIterator groupShardsIterator = clusterService.operationRouting().searchShards(clusterService.state(), indexNameExpressionResolver.concreteIndices(clusterService.state(), request), null, null);
+        final AtomicReferenceArray<ReplicatedRefreshResponse> shardsResponses = new AtomicReferenceArray(groupShardsIterator.size());
+        final AtomicInteger numPrimariesToProcess = new AtomicInteger(groupShardsIterator.size());
```

Comment:
we have a `CountDown` class that helps to do this and prevents you from counting down more than actually specified...